app/lib/service/download_counts/download_counts.dart

```diff
@@ -0,0 +1,152 @@
+  DateTime? lastDate;
+
+  /// A list of tuples containing a string describing a major `versionRange`,
+  /// e.g. '>=1.0.0-0 <2.0.0', and a list of integers with a count for each day.
+  /// The 'counts' list contains at most [maxAge] entries. The first entry in
+  /// represents the number of downloads on `lastDate` followed by the downloads
+  ///  on `lastDate` - 1 and so on. E.g.
+  ///
+  ///  counts = [ 42, 21, 55 ]
+  ///              ▲   ▲   ▲
+  ///              │   │   └──────────── Download count on lastDate - 2 days
+  ///              │   │
+  ///              │   └──────────────── Download count on lastDate - 1 day
+  ///              │
+  ///              └──────────────────── Download count on lastDate
+  ///
+  ///
+  /// [majorCounts] has at most [maxRanges] elements and is sorted by version
+  ///  ranges.
+  final majorCounts = <({String versionRange, List<int> counts})>[];
+  final minorCounts = <({String versionRange, List<int> counts})>[];
+  final patchCounts = <({String versionRange, List<int> counts})>[];
+  final totalCounts = <int>[];
+
+  CountData();
+
+  /// Process and store download counts given in [dayCounts] on the form
+  /// {version: #downloads} for a date given by [dateTime].
+  void addDownloadCounts(Map<String, int> dayCounts, DateTime dateTime) {
+    final date = DateTime(dateTime.year, dateTime.month, dateTime.day);
+    lastDate ??= date.addCalendarDays(-1);
+
+    final nextLastDate = date.isAfter(lastDate!) ? date : lastDate!;
+    final countsIndex = nextLastDate.difference(date).inDays;
+
+    if (countsIndex >= maxAge) {
+      // We don't store counts that are more than two years old.
+      return;
+    }
+
+    // TODO(zarah): call this with other ranges
+    _prepareDates(date, lastDate!, countsIndex, majorCounts);
+
+    _processCounts(
+        dayCounts, countsIndex, majorCounts, _createNewMajorVersionRange);
+
+    lastDate = nextLastDate;
+  }
+
+  void _prepareDates(
+    DateTime date,
+    DateTime workingLastDate,
+    int countsIndex,
+    List<({List<int> counts, String versionRange})> versionCounts,
+  ) {
+    if (date.isAfter(workingLastDate)) {
+      final zerosList = List.filled(date.difference(workingLastDate).inDays, 0);
+      for (int i = 0; i < versionCounts.length; i++) {
+        // Fill in with 0 on days with no data.
+        final newCounts =
+            [...zerosList, ...versionCounts[i].counts].take(maxAge).toList();
+        versionCounts[i] =
+            (counts: newCounts, versionRange: versionCounts[i].versionRange);
+      }
+    } else if (date.isAtOrBefore(workingLastDate)) {
+      versionCounts.forEach((versionCount) {
+        if (versionCount.counts.length < countsIndex + 1) {
+          // Fill in with 0 on older days with no data.
+          versionCount.counts.addAll(
+              List.filled(countsIndex - versionCount.counts.length + 1, 0));
+        }
+
+        // Reset the counts for this date.
+        versionCount.counts[countsIndex] = 0;
```

Comment:
Not sure I'm reading this right. Should this not be done in an `else` branch? Such that you have:
```
if (date.isAfter(workingLastDate)) {
  // ... prepend zeros ...
} else if (date.isAtOrBefore(workingLastDate)) {
  // ... append zeros ...
} else {
  // reset counts for this date.
}
```

Reply:
It shouldn't as discussed.